### PR TITLE
feat: add hmmsearch module and tests

### LINKED ISSUE #930: new module: hmmer/hmmsearch
I think it would be good to have a module for [hmmer/hmmsearch](https://bioconda.github.io/recipes/hmmer/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Midnighter**
## PR checklist

Closes #930

- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/hmmer/hmmsearch/meta.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+name: hmmer_hmmsearch
+description: search profile(s) against a sequence database
+keywords:
+  - hidden Markov model
```
Comment (jfy133) [inline modules/hmmer/hmmsearch/meta.yml:4]:
```suggestion
  - hidden markov model```

File: modules/hmmer/hmmsearch/meta.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - hmmfile:
+      type: file
+      description: One or more HMM profiles created with hmmbuild
+      pattern: "*.{hmm}"
+  - seqdb:
+      type: file
+      description: Database of sequences
```
Comment (jfy133) [inline modules/hmmer/hmmsearch/meta.yml:29]:
```suggestion
      description: Database of sequences in FASTA format
```

File: modules/hmmer/hmmsearch/meta.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      pattern: "*.{fasta,fna,faa,fa}"
+  - write_align:
+      type: val
+      description: Flag to write optional alignment output
```
Comment (jfy133) [inline modules/hmmer/hmmsearch/meta.yml:33]:
```suggestion
      description: Flag to write optional alignment output. Specify with 'true' to output
```

File: modules/hmmer/hmmsearch/meta.yml (hunk 1 above)
Comment (Midnighter) [reply modules/hmmer/hmmsearch/meta.yml:4]: Normally I would capitalize Markov since it's a given name but I guess keywords should be all lowercase?

**Comment by erikrikarddaniel** [comment]
> Overall LGTM, few minor docs tweaks
> 
> The only thing we are not complying with is compressed files. I just discovered `hmmer` can accept gzip files (http://eddylab.org/software/hmmer/Userguide.pdf - page 208). So theoretically we should update this and also maybe update `hmmalign`, but we should get feedback from @erikrikarddaniel whether this would be an issue or not.

The only way I've handled gzipped input to `hmmsearch` is with `<(gunzip -c file.gz)`. From the documentation you linked to, I only see it mentioned for a formatting tool in hmmer, `esl-translate`.

In principle, I'm all for adding the possibility to read gzipped data to the modules. I'm a bit short of time the coming few weeks though.

**Comment by jfy133** [comment]
@erikrikarddaniel OK if you'r eOK with it in principle maybe @Midnighter and I can test next week and update both modules. I just wanted to check we weren't going to break anything for you downstream.

**Comment by erikrikarddaniel** [comment]
> @erikrikarddaniel OK if you'r eOK with it in principle maybe @Midnighter and I can test next week and update both modules. I just wanted to check we weren't going to break anything for you downstream.

Absolutely, I'd welcome this, particularly for the search module, as it's a common use case for me too.

File: tests/modules/hmmer/hmmsearch/test.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  command: nextflow run tests/modules/hmmer/hmmsearch -entry test_hmmer_hmmsearch_optional -c tests/config/nextflow.config
+  tags:
+    - hmmer/hmmsearch
+    - hmmer
+  files:
+    - path: output/hmmer/test.sto.gz
+      md5sum: d3121aa33455074c566fb7f8fdcda7b0
+    - path: output/hmmer/test.domtbl.gz
+      contains:
+        - '# [ok]'
```
[the remark this answers is not in the document]
Comment (Midnighter) [reply tests/modules/hmmer/hmmsearch/test.yml:23]: Apparently, it does, yes.

File: modules/hmmer/hmmsearch/meta.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - output:
+      type: file
+      description: Human readable output summarizing hmmsearch results
```
[the remark this answers is not in the document]
Comment (Midnighter) [reply modules/hmmer/hmmsearch/meta.yml:53]: No idea, is there a hmmer module for MultiQC?

**Comment by jfy133** [comment]
No there isn't so we should be good 😎

https://multiqc.info/#supported-tools

---

**Pull request merged**